storage/diemdb/src/event_store/mod.rs

```diff
@@ -238,6 +282,84 @@ impl EventStore {
             })
             .collect::<Result<Vec<_>>>()
     }
+
+    /// Finds the first event sequence number in a specified stream on which `comp` returns false.
+    /// (assuming the whole stream is partitioned by `comp`)
+    fn search_for_event_lower_bound<C>(
+        &self,
+        event_key: &EventKey,
+        mut comp: C,
+        ledger_version: Version,
+    ) -> Result<Option<u64>>
+    where
+        C: FnMut(&ContractEvent) -> Result<bool>,
+    {
+        let mut begin = 0u64;
+        let mut end = match self.get_latest_sequence_number(ledger_version, event_key)? {
+            Some(s) => s
+                .checked_add(1)
+                .ok_or_else(|| format_err!("event sequence number overflew."))?,
+            None => return Ok(None),
+        };
+
+        // overflow not possible
+        #[allow(clippy::integer_arithmetic)]
+        {
+            let mut count = end - begin;
+            while count > 0 {
+                let step = count / 2;
+                let mid = begin + step;
+                let event = self.get_event_by_key(event_key, mid, ledger_version)?;
+                if comp(&event)? {
+                    begin = mid + 1;
+                    count -= step + 1;
+                } else {
+                    count = step;
+                }
+            }
+        }
```

Comment:
oh my... coding interviews...

aren't we supposed to break if we find it? shouldn't this be a > go up, < go down, == stop.

Reply:
Yeah.. you need to distinguish `lower_bound`, `upper_bound`, `equal_range` and the toy exact binary search found on the internet that returns a random index among duplications. Great preparation to do before a C++ interview (I made that up).